Share queue stats among all apps on a NIC

takikawa opened the pull request:
This PR revises queue stats handling by providing all the queue stats in each process' `pci` shm directory. It does this by symlinking all of them to a "global" shm sub-directory in the `/intel-mp` directory.

This doesn't quite solve the `snabb top` issues yet because I think symlinks are not yet followed by some layer that `snabb top` uses. Once `snabb top` follows the symlinks it should show all the counters properly. I'm looking into fixing it, but if anyone has any ideas on where to find the cause (maybe at the inotify level?) it would be helpful.

Also, this exposes the `rxcounter` and `txcounter` values in an extra counter, which could be used to filter the queue stats shown (since there may be non-zero queue counters of no interest for a given process). This part is WIP but could be useful if `snabb top` actually looks at it.

takikawa committed:
intel_mp: store shm stats in global directory

Instead of storing intel_mp stats in a per-process stats
directory, store it in the global intel_mp directory that's
indexed by PCI address (the same place where pool number
allocations are tracked).

This lets intel_mp apps access counters updated by the main
process even if they live in different processes. It also makes
it easier to track queue counters in `snabb top`

Also exposes the rxcounter/txcounter assignment in the per-process
shm area.

takikawa committed:
Go back to avoiding setting run_stats

The run_stats flag should be used to specify which app
is in charge of syncing stats. With commit
59f2f8a76c607eb851b7bce52ea85bcf53ea1fa7 it is unnecessary
to set this for the lwAFTR.

dpino commented:
Nice work. It looks good to me so far.

I also think `snabb top` issues are at inotify level. I would need to look into it to actually understand what's the issue.

dpino commented:
It seems like inotify doesn't report events on symbolic links. Here's an example using user-space tool `inotifywait`:

```bash
$ mkdir temp
$ inotifywait -m temp          
Setting up watches.                         
Watches established.                        
temp/ CREATE,ISDIR foo
```
```
temp $ touch foo
```

But if the observed file is a symbolic link to a directory

```bash
$ ln -s temp tmp
$ inotifywait -m tmp
Setting up watches.
Watches established.
```

```bash
temp $ touch bar
```

Nothing happens.

takikawa committed:
inotify: adjust is_dir to allow symlinks to dirs

This makes `snabb top` work on symlinked pci directories
for intel_mp stats

takikawa commented:
Update: I think I've solved the inotify issue. It was just a matter of changing how the snabb library defines what a directory is for this case. The queue stats now show up again.

I'll see if I can make it filter based on the actually enabled counters.

takikawa committed:
Filter queue counters based on enabled bitmap

This filters out queue counters that are unused by a particular
process in its tree view, even if the NIC it's using has traffic
on those queues from another process

Also adjusts some tests to use rx/txcounter in order to show
stats separately in `snabb top` while tests run

takikawa commented:
I pushed another commit that filters out irrelevant queue counters in `snabb top`. This change was more complicated than I thought and to be honest I'm not that happy with it, but it works and passes the `intel_mp` tests. I think it could use more testing though.

The complication comes from how the `rxcounters` / `txcounters` counters work. They're counters that store a bitmap indicating which queue counters are used by a process. Since this is pid-specific, it shouldn't go in the shared shm folder. This means we can't symlink the whole folder, and have to symlink each individual shared counter instead (otherwise the writes would go to the same counter).

I guess an alternative approach would be to add another folder inside the shared shm folder that's indexed by pid, but then the folder structure looks a little silly `8251/pci/02:00.0/8251/rxcounters.counter` with the repeated pid. Or the counter could be stored in a path entirely separate from the stats counters, like `8251/intel_mp_config/02:00.0/rxcounters.counter`. I'm starting to think the latter approach with a new folder might be better now, so I will try it too.

takikawa committed:
intel_mp: add a new test for rxcounters & workers

takikawa committed:
snabb top: refactor queue stats filtering

Use a simpler method to figure out which queue stats to filter
out. Revert change to symlink each counter for intel_mp and
just symlink the whole folder again. Then add a counter in the
per-app shm that just stores the rxcounter/txcounter.

Snabb top then looks through all the apps in a process and filters
out queue counters that don't match any app. This is not totally precise
in the case that a process uses multiple NICs with counters set
(it may show more counters than necessary).

dpino committed:
Fix error while opening rrd.file

dpino commented:
Related with this PR https://github.com/takikawa/snabb/pull/4

takikawa committed:
snabb top: fix worker nesting with shared pci

Instead of changing inotify to crawl links too, we will treat
the pci link specially in snabb top instead. This also reverts
commit 0d641de05832facc16ba368d8b982be33ae1ecd0 which made the
inotify change.

dpino committed:
Show rxbits and txbits instead of bytes
